probe-rs/src/lib.rs

```diff
@@ -101,8 +107,291 @@ pub use crate::core::{
 };
 pub use crate::error::Error;
 pub use crate::memory::MemoryInterface;
-pub use crate::probe::{
-    list::Lister, AttachMethod, DebugProbe, DebugProbeError, DebugProbeInfo, DebugProbeSelector,
-    Probe, ProbeCreationError, ProbeDriver, WireProtocol,
-};
+use crate::probe::{AttachMethod, DebugProbe, DebugProbeError, WireProtocol};
 pub use crate::session::{Permissions, Session};
+
+/// The Probe struct is a generic wrapper over the different
+/// probes supported.
+///
+/// # Examples
+///
+/// ## Open the first probe found
+///
+/// The `list_all` and `from_probe_info` functions can be used
+/// to create a new `Probe`:
+///
+/// ```no_run
+/// use probe_rs::{probe::list::Lister, Probe};
+///
+/// let lister = Lister::new();
+///
+/// let probe_list = lister.list_all();
+/// let probe = probe_list[0].open(&lister);
+/// ```
+#[derive(Debug)]
+pub struct Probe {
```

Comment:
Why do we not have this in the `probe-rs/src/probe.rs` file anymore? I think moving this is wrong. Is there any practical reason I do not get? I think it's valuable to work around it. The code structure would be different than one would expect if we change this.

Reply:
> The code structure would be different than one would expect if we change this.

I absolutely hate import structures that don't match the crate's structure :) And Dario really wants me to call the (DebugProbe) trait Probe, too, which lives in the same module so it can't be defined currently. I don't really care either way (regarding the trait's name), but this guy is exported as probe_rs::Probe, not probe_rs::probe::Probe so it shouldn't live in `probe`.

Reply:
I don't think stuffing everything into lib.rs is a good thing because it will just flood the file. I think for that use case reexports are fine.
Also, the double naming is a pain. Maybe we should call the trait ProbeInterface like we do for some other things. Ultimately idk what's best :/

Reply:
Best is to not reexport everything from root :P I'll undo this part asap :)

Reply:
Although now if `probe` is public, `Probe` shouldn't be reexported anyway as it'll be available through two paths 🤔

Reply:
Hmm good question. Do we need it in the root? I mean I initially did that. But nowadays with r-a and intellij, it's not that big of a deal anymore imo.

Reply:
> Do we need it in the root? I mean I initially did that. But nowadays with r-a and intellij, it's not that big of a deal anymore imo.

No but moving it out means I have to update like 70 files because everything is reexported from root.